// @dotmac/http-client/src/generated-schema.ts
/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {
  "/auth/token": {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get?: never;
    put?: never;
    /** Issue access and refresh tokens */
    post: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody: {
        content: {
          "application/json": components["schemas"]["TokenRequest"];
        };
      };
      responses: {
        /** @description Token pair issued successfully */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["TokenResponse"];
          };
        };
        /** @description Invalid credentials supplied */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["ErrorResponse"];
          };
        };
      };
    };
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  "/auth/token/verify": {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get?: never;
    put?: never;
    /** Verify access token validity */
    post: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody: {
        content: {
          "application/json": components["schemas"]["VerifyTokenRequest"];
        };
      };
      responses: {
        /** @description Token claims */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["TokenClaims"];
          };
        };
        /** @description Token is invalid or expired */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["ErrorResponse"];
          };
        };
      };
    };
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  "/auth/sessions": {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    /** List active sessions */
    get: {
      parameters: {
        query?: {
          /** @description Filter sessions by user identifier */
          user_id?: string;
          /** @description If true only active sessions are returned */
          active_only?: boolean;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description Collection of sessions */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["SessionListResponse"];
          };
        };
      };
    };
    put?: never;
    /** Create a new session */
    post: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody: {
        content: {
          "application/json": components["schemas"]["CreateSessionRequest"];
        };
      };
      responses: {
        /** @description Session created */
        201: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["SessionResponse"];
          };
        };
        /** @description Invalid payload */
        400: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["ErrorResponse"];
          };
        };
      };
    };
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  "/auth/sessions/{session_id}": {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    /** Fetch a session by identifier */
    get: {
      parameters: {
        query?: never;
        header?: never;
        path: {
          /** @description Unique session identifier */
          session_id: components["parameters"]["SessionIdParam"];
        };
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description Session details */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["SessionResponse"];
          };
        };
        /** @description Session not found */
        404: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["ErrorResponse"];
          };
        };
      };
    };
    put?: never;
    post?: never;
    /** Revoke a session */
    delete: {
      parameters: {
        query?: never;
        header?: never;
        path: {
          /** @description Unique session identifier */
          session_id: components["parameters"]["SessionIdParam"];
        };
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description Session revoked successfully */
        204: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Session not found */
        404: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["ErrorResponse"];
          };
        };
      };
    };
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  "/auth/api-keys": {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    /** List API keys */
    get: {
      parameters: {
        query?: {
          /** @description Filter to include only active keys */
          active_only?: boolean;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description API key summaries */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["ApiKeyListResponse"];
          };
        };
      };
    };
    put?: never;
    /** Create an API key */
    post: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody: {
        content: {
          "application/json": components["schemas"]["CreateApiKeyRequest"];
        };
      };
      responses: {
        /** @description API key created */
        201: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["ApiKeyResponse"];
          };
        };
      };
    };
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  "/auth/api-keys/{key_id}": {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get?: never;
    put?: never;
    post?: never;
    /** Revoke an API key */
    delete: {
      parameters: {
        query?: never;
        header?: never;
        path: {
          /** @description Unique API key identifier */
          key_id: components["parameters"]["ApiKeyIdParam"];
        };
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description API key revoked */
        204: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description API key not found */
        404: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["ErrorResponse"];
          };
        };
      };
    };
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  "/health": {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    /** Platform health check */
    get: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description Health information */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["HealthResponse"];
          };
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  "/api/platform/v1/admin/deployment/stats": {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    /** Get deployment statistics summary */
    get: {
      parameters: {
        query?: {
          /** @description Filter statistics for a specific tenant */
          tenant_id?: number;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description Aggregated deployment statistics */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["DeploymentStatsResponse"];
          };
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  "/api/platform/v1/admin/deployment/stats/templates": {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    /** Get deployment template usage metrics */
    get: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description Template usage summary */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["DeploymentTemplateUsageResponse"];
          };
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  "/api/platform/v1/admin/deployment/stats/resources": {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    /** Get resource allocation totals */
    get: {
      parameters: {
        query?: {
          /** @description Filter allocation totals for a specific tenant */
          tenant_id?: number;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description Aggregated resource allocation totals */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            "application/json": components["schemas"]["DeploymentResourceAllocationResponse"];
          };
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
}
export type webhooks = Record<string, never>;
export interface components {
  schemas: {
    TokenRequest: {
      /** @description Username or email */
      username: string;
      /** Format: password */
      password: string;
      tenant_id?: string | null;
      /** @default [] */
      scopes: string[];
    };
    VerifyTokenRequest: {
      /** @description Access token to verify */
      token: string;
    };
    TokenResponse: {
      access_token: string;
      refresh_token?: string | null;
      /** @enum {string} */
      token_type: "Bearer";
      /** Format: int32 */
      expires_in: number;
    };
    TokenClaims: {
      /** @description Subject identifier */
      sub: string;
      /** @enum {string} */
      type: "access" | "refresh";
      /** @description Token issuer */
      iss: string;
      /** @description Token audience */
      aud: string;
      /** @description Space delimited scopes */
      scope?: string;
      scopes?: string[];
      tenant_id?: string | null;
      /** Format: int64 */
      exp?: number;
      /** Format: int64 */
      iat?: number;
      jti?: string;
    };
    ErrorResponse: {
      detail?: string;
      error?: string;
      code?: string;
    };
    CreateSessionRequest: {
      user_id: string;
      tenant_id?: string | null;
      metadata?: {
        [key: string]: unknown;
      };
    };
    SessionResponse: {
      session_id: string;
      user_id: string;
      tenant_id: string | null;
      /** Format: date-time */
      created_at: string;
      /** Format: date-time */
      last_accessed?: string | null;
      /** Format: date-time */
      expires_at: string;
      /** @enum {string} */
      status: "active" | "expired" | "revoked";
      metadata: {
        [key: string]: unknown;
      };
    };
    SessionListResponse: {
      sessions: components["schemas"]["SessionResponse"][];
      /** Format: int32 */
      total: number;
    };
    HealthResponse: {
      /** @enum {string} */
      status: "healthy" | "degraded" | "unhealthy";
      /** Format: date-time */
      timestamp: string;
      checks?: {
        [key: string]: components["schemas"]["HealthComponentStatus"];
      };
      version?: string;
    };
    HealthComponentStatus: {
      /** @enum {string} */
      status: "healthy" | "unhealthy";
      /** Format: float */
      response_time?: number;
    };
    CreateApiKeyRequest: {
      name: string;
      /** @default [] */
      scopes: string[];
      /** Format: date-time */
      expires_at?: string | null;
    };
    ApiKeyResponse: {
      key_id: string;
      api_key: string;
      name: string;
      scopes: string[];
      /** Format: date-time */
      created_at: string;
      /** Format: date-time */
      expires_at?: string | null;
    };
    ApiKeySummary: {
      key_id: string;
      name: string;
      /** Format: date-time */
      created_at: string;
      /** Format: date-time */
      expires_at?: string | null;
    };
    ApiKeyListResponse: {
      api_keys: components["schemas"]["ApiKeySummary"][];
      /** Format: int32 */
      total: number;
    };
    DeploymentStatsResponse: {
      /** Format: int64 */
      total_instances: number;
      states: {
        /** Format: int64 */
        active: number;
        /** Format: int64 */
        provisioning: number;
        /** Format: int64 */
        failed: number;
        /** Format: int64 */
        suspended: number;
      };
      health: {
        /** Format: int64 */
        healthy: number;
        /** Format: int64 */
        degraded: number;
        /** Format: int64 */
        unhealthy: number;
      };
    };
    DeploymentTemplateUsageEntry: {
      template_name: string;
      display_name: string;
      /** Format: int64 */
      instances: number;
    };
    DeploymentTemplateUsageResponse: {
      templates: components["schemas"]["DeploymentTemplateUsageEntry"][];
      /** Format: int64 */
      total_templates: number;
      /** Format: int64 */
      total_instances: number;
    };
    DeploymentResourceAllocationResponse: {
      /**
       * Format: int64
       * @description Total allocated CPU cores
       */
      total_cpu: number;
      /**
       * Format: int64
       * @description Total allocated memory in GiB
       */
      total_memory: number;
      /**
       * Format: int64
       * @description Total allocated storage in GiB
       */
      total_storage: number;
    };
  };
  responses: never;
  parameters: {
    /** @description Unique session identifier */
    SessionIdParam: string;
    /** @description Unique API key identifier */
    ApiKeyIdParam: string;
  };
  requestBodies: never;
  headers: never;
  pathItems: never;
}
export type $defs = Record<string, never>;
export type operations = Record<string, never>;
